File: scripts/query-session.mjs
// Quick MongoDB query to inspect the most recent agent session's requests
import { MongoClient } from "mongodb";
import { MONGO_URI, MONGO_DB_NAME } from "../secrets.js";
import { COLLECTIONS } from "../src/constants.js";

const client = new MongoClient(MONGO_URI);
await client.connect();
const db = client.db(MONGO_DB_NAME);

// Find the most recent session
const sessions = await db.collection(COLLECTIONS.AGENT_SESSIONS)
  .find()
  .sort({ updatedAt: -1 })
  .limit(1)
  .toArray();

if (!sessions.length) {
  console.log("No sessions found");
  process.exit(1);
}

const session = sessions[0];
// The session's `id` field is the UUID used as agentSessionId in requests
const sid = session.id || session._id;
console.log(`Session _id: ${session._id}`);
console.log(`Session id:  ${session.id}`);
console.log(`Title: ${(session.title || "?").substring(0, 80)}`);
console.log(`Project: ${session.project}`);
console.log(`Username: ${session.username}`);
console.log();

// Discover all descendant session IDs (same traversal as agent-sessions.js)
const allSessionIds = new Set([sid]);
let frontier = [sid];
for (let depth = 0; depth < 10 && frontier.length > 0; depth++) {
  const childIds = await db.collection(COLLECTIONS.REQUESTS)
    .distinct("agentSessionId", {
      parentAgentSessionId: { $in: frontier },
      agentSessionId: { $nin: [...allSessionIds] },
    });
  if (childIds.length === 0) break;
  const newIds = childIds.filter(Boolean);
  for (const id of newIds) allSessionIds.add(id);
  frontier = newIds;
}

console.log(`Total session IDs: ${allSessionIds.size} (main + ${allSessionIds.size - 1} workers)`);
if (allSessionIds.size > 1) {
  console.log(`Worker IDs: ${[...allSessionIds].filter(id => id !== sid).join(", ")}`);
}
console.log();

// Get all requests
const requests = await db.collection(COLLECTIONS.REQUESTS)
  .find({ agentSessionId: { $in: [...allSessionIds] } })
  .sort({ timestamp: 1 })
  .toArray();

console.log(`${"=".repeat(130)}`);
console.log(`  All ${requests.length} requests for session`);
console.log(`${"=".repeat(130)}`);

for (let i = 0; i < requests.length; i++) {
  const r = requests[i];
  const op = (r.operation || r.endpoint || "?").padEnd(25);
  const model = (r.model || "?").substring(0, 30).padEnd(30);
  const inp = (r.inputTokens || 0);
  const out = (r.outputTokens || 0);
  const ts = r.timestamp || "?";
  const isWorker = r.agentSessionId !== sid;
  const marker = isWorker ? " [WORKER]" : "";
  console.log(`  ${String(i + 1).padStart(2)}. ${op} | ${model} | in:${String(inp).padStart(8)} out:${String(out).padStart(8)}${marker}  ${ts}`);
}

// Summary by operation
const byOp = {};
for (const r of requests) {
  const op = r.operation || r.endpoint || "?";
  const isWorker = r.agentSessionId !== sid;
  const key = `${op}${isWorker ? " [WORKER]" : ""}`;
  if (!byOp[key]) byOp[key] = { count: 0, inputTokens: 0, outputTokens: 0 };
  byOp[key].count++;
  byOp[key].inputTokens += r.inputTokens || 0;
  byOp[key].outputTokens += r.outputTokens || 0;
}

console.log(`\n${"=".repeat(80)}`);
console.log("  Summary by operation");
console.log(`${"=".repeat(80)}`);
for (const [op, data] of Object.entries(byOp).sort((a, b) => b[1].count - a[1].count)) {
  console.log(`  ${op.padEnd(40)} | ${String(data.count).padStart(3)} requests | in:${String(data.inputTokens).padStart(10)} out:${String(data.outputTokens).padStart(10)}`);
}

await client.close();

File: scripts/query-session.py
#!/usr/bin/env python3
"""Query the most recent agent session and list all its requests."""
import json
import subprocess
import sys

# Fetch the most recent session
result = subprocess.run(
    ["curl", "-s", "http://localhost:7777/agent-sessions?limit=1", "-H", "x-project: default"],
    capture_output=True, text=True
)

data = json.loads(result.stdout)
# Could be list or { sessions: [...] }
sessions = data if isinstance(data, list) else data.get("sessions", [])
if not sessions:
    print("No sessions found")
    sys.exit(1)

session = sessions[0]
sid = session["_id"]
title = session.get("title", "?")
print(f"Session: {sid}")
print(f"Title: {title[:80]}")
print(f"Created: {session.get('createdAt', '?')}")
print()

# Fetch full session with stats
result2 = subprocess.run(
    ["curl", "-s", f"http://localhost:7777/agent-sessions/{sid}", "-H", "x-project: default"],
    capture_output=True, text=True
)

full = json.loads(result2.stdout)
st = full.get("stats", {})
if st:
    print(f"Total Requests: {st.get('requestCount', 0)}")
    print(f"Models: {st.get('models', [])}")
    print(f"Input Tokens: {st.get('totalInputTokens', 0):,}")
    print(f"Output Tokens: {st.get('totalOutputTokens', 0):,}")
    print(f"Total Tokens: {st.get('totalTokens', 0):,}")
    print(f"Total Cost: ${st.get('totalCost', 0):.5f}")
    
    orch = st.get("orchestrator", {})
    work = st.get("workers", {})
    if orch:
        print(f"\nOrchestrator: {orch.get('requestCount', 0)} requests, out={orch.get('totalOutputTokens',0):,}")
    if work:
        print(f"Workers: {work.get('requestCount', 0)} requests, out={work.get('totalOutputTokens',0):,}")
    print()

# Query requests directly from MongoDB via admin stats endpoint
# Try the requests listing endpoint
result3 = subprocess.run(
    ["curl", "-s", f"http://localhost:7777/admin/stats/requests?agentSessionId={sid}&limit=100", "-H", "x-project: default"],
    capture_output=True, text=True
)

try:
    req_data = json.loads(result3.stdout)
except:
    # Try alternate endpoint
    result3 = subprocess.run(
        ["curl", "-s", f"http://localhost:7777/admin/requests?agentSessionId={sid}&limit=100", "-H", "x-project: default"],
        capture_output=True, text=True
    )
    try:
        req_data = json.loads(result3.stdout)
    except:
        print(f"Could not query requests. Response: {result3.stdout[:300]}")
        sys.exit(1)

requests_list = req_data if isinstance(req_data, list) else req_data.get("requests", req_data.get("data", []))

if not requests_list:
    print(f"No requests found via admin route.")
    print(f"Response preview: {json.dumps(req_data)[:300]}")
else:
    print(f"{'='*100}")
    print(f"  All {len(requests_list)} requests for session")
    print(f"{'='*100}")
    for i, r in enumerate(requests_list):
        op = r.get("operation", r.get("endpoint", "?"))
        model = (r.get("model", "?") or "?")[:30]
        inp = r.get("inputTokens", 0) or 0
        out = r.get("outputTokens", 0) or 0
        ts = r.get("timestamp", "?")
        asid = r.get("agentSessionId", "?")
        is_worker = asid != sid
        marker = " [WORKER]" if is_worker else ""
        print(f"  {i+1:2d}. {op:25s} | {model:30s} | in:{inp:>8,} out:{out:>8,}{marker}  {ts}")
